File: 双指针/15.三数之和.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        # 方法：先确定一个数，在它之后的范围中再找两个数，使得三者的和为0

        result = []
        nums.sort()  # 先对数组进行排序

        for i in range(len(nums) - 2):
            # 如果后面的数字和前面的数字重复就跳过这些数字, 
            # 跳过重复的数字 注意防止越界 if nums[i] == nums[i + 1]:
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            left, right = i + 1, len(nums) - 1
            while left < right:
                _sum = nums[i] + nums[left] + nums[right]
                if _sum < 0:
                    left += 1
                elif _sum > 0:
                    right -= 1
                else:
                    result.append([nums[i], nums[left], nums[right]])
                    # 跳过重复数字
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                    left += 1
                    right -= 1
        return result

File: 双指针/15.三数之和.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
  const res = [];
  nums.sort((a, b) => a - b);
  for (let i = 0; i < nums.length - 2; i++) {
    //! 最后空两个
    if (i > 0 && nums[i] === nums[i - 1]) continue; //! 和前一个比较

    let j = i + 1,
      k = nums.length - 1; //! 从i+1开始
    while (j < k) {
      //! 注意两个指针肯定不应该指向同一个数字
      const sum = nums[i] + nums[j] + nums[k];
      if (sum > 0) {
        k--;
      } else if (sum < 0) {
        j++;
      } else {
        res.push([nums[i], nums[j], nums[k]]);
        while (j < k && nums[j] === nums[j + 1]) j++;
        while (j < k && nums[k] === nums[k - 1]) k--;
        //! 此时j和j + 1  k和k - 1不相等
        j++;
        k--;
      }
    }
  }
  return res;
};
